### PR TITLE
[P2T support No.2-3、14、26、28] paddle.autograd.[hessian、jacobian]、paddle.incubate.nn.functional.[block_multihead_attention、masked_multihead_attention、variable_length_memory_efficient_attention]

### PR DESCRIPTION
**Pull request by cangtianhuang**
### 1. [paddle.autograd.hessian](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/autograd/hessian_cn.html#hessian)

无对应 api，在 mapping 中注册 HessianRule，HessianRule 实现了返回用于计算海森矩阵的 Hessian 实例，具体见 Rule 代码。但是 accuracy 暂不支持实例的精度比较，引擎暂不需要补齐此功能，因此未进行进度比较

### 2. [paddle.autograd.jacobian](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/autograd/jacobian_cn.html#jacobian)

无对应 api，在 mapping 中注册 JacobianRule，JacobianRule 实现了返回用于计算雅可比矩阵的 Jacobian 实例，具体见 Rule 代码。未进行进度比较

### 3. [paddle.incubate.nn.functional.block_multihead_attention](https://www.paddlepaddle.org.cn/documentation/docs/en/develop/api/paddle/incubate/nn/functional/block_multihead_attention_en.html#block-multihead-attention)

无对应 api，此 api 的参数过于复杂，决定不支持转换

### 4. [paddle.incubate.nn.functional.masked_multihead_attention](https://www.paddlepaddle.org.cn/documentation/docs/en/develop/api/paddle/incubate/nn/functional/masked_multihead_attention_en.html#masked-multihead-attention)

无对应 api，此 api 的参数较为复杂，在 mapping 中注册 MaskedMultiheadAttentionRule，尝试转换后仍无法对齐行为，决定不支持转换

### 5. [paddle.incubate.nn.functional.variable_length_memory_efficient_attention](https://www.paddlepaddle.org.cn/documentation/docs/en/develop/api/paddle/incubate/nn/functional/variable_length_memory_efficient_attention_en.html#variable-length-memory-efficient-attention)

无对应 api，此 api 的参数较为复杂，在 mapping 中注册 VariableLengthMemoryEfficientAttentionRule，反复尝试后初步对齐行为

## 其他修改

exec_globals 中增加了 math import，使得 rule 中可以使用 math 相关操作

## 测试结果

paddle.autograd.[hessian jacobian]、paddle.incubate.nn.functional.[block_multihead_attention masked_multihead_attention] 的所有配置移入 paddleonly_1；paddle.incubate.nn.functional.variable_length_memory_efficient_attention 部分有 accuracy 精度问题，移入 accuracy_error_1 并在 fresh_report 中记录，其余配置移入 support2torch_1：

image

- https://github.com/PaddlePaddle/Paddle/issues/72643

### CODE REVIEW

File: tester/paddle_to_torch/rules.py [hunk 1]
```diff
@@ -3211,6 +3486,70 @@ def apply(self, paddle_api: str) -> ConvertResult:
...
+    key_seq_len = key.shape[2]
+    # Broadcast key and value to match query's num_heads if needed
+    if key.shape[1] != num_heads:
+        # Repeat key and value along the num_heads dimension
+        repeat_factor = num_heads // key.shape[1]
+        key = key.repeat(1, repeat_factor, 1, 1)
+        value = value.repeat(1, repeat_factor, 1, 1)
+    # Default scale if not provided
+    if scale is None:
+        scale = math.sqrt(1.0 / head_size)
```
Comment (Cutelemon6) [inline tester/paddle_to_torch/rules.py:3513]: 可以使用指数运算符  (1.0 / hs) ** 0.5 来避免使用 math.sqrt 和 引入 math 库

File: tester/paddle_to_torch/rules.py [hunk 2]
```diff
@@ -1929,6 +2084,134 @@ def apply(self, paddle_api: str) -> ConvertResult:
...
+        cache_kvs_out = torch.stack((k, v), dim=0)
+    else:
+        k = k.unsqueeze(-2)  # [batch_size, num_head, 1, head_dim]
+        v = v.unsqueeze(-2)
+        cache_kvs_out = torch.stack((k, v), dim=0)
+    # Reshape for attention: [batch_size, num_head, seq_len, head_dim]
+    q = q.unsqueeze(2)  # [batch_size, num_head, 1, head_dim]
+    seq_len_kv = k.shape[2]
+    # Compute attention scores
+    attn_scores = torch.matmul(q, k.transpose(-2, -1)) / math.sqrt(head_dim)  # [batch_size, num_head, 1, seq_len_kv]
```
Comment (Cutelemon6) [inline tester/paddle_to_torch/rules.py:2165]: 可以使用指数运算符 num ** 0.5 来避免使用 math.sqrt 和 引入 math 库

File: tester/accuracy.py [hunk 3]
```diff
@@ -61,7 +63,7 @@ def test(self):
             # 以下代码等价于:
             # torch_output = Paddle2TorchConverter.execute(convert_result, self.torch_args, self.torch_kwargs)
             # 准备执行环境，将参数(torch tensors)直接映射至locals
-            exec_globals = {"torch": torch}
+            exec_globals = {"torch": torch, "math": math}
```
Comment (Cutelemon6) [inline tester/accuracy.py:66]: 避免引入 math 库

File: tester/paddle_to_torch/rules.py [hunk 4]
```diff
@@ -1638,6 +1631,80 @@ def taylor(M, nbar=4, sll=30, norm=True, sym=True, dtype=torch.float64):
...
+            result = self._compute_hessian_single(self.xs[idx1], batch_idx)
+        else:
+            result = self._compute_hessian_tuple(idx1, idx2, batch_idx)
+        # 缓存结果
+        self.cache[cache_key] = result
+        return result
+"""
+        core = "result = Hessian(ys=ys, xs=xs, batch_axis=batch_axis)"
+        code = Code(preprocess=pre.splitlines(), core=[core])
+        return ConvertResult.success(paddle_api, code, is_torch_corresponding=False)
```
Comment (Cutelemon6) [inline tester/paddle_to_torch/rules.py:1705]: 在 accuracy.py 中添加 torch Hessian 转换到 torch Tensor

File: tester/accuracy.py (hunk 3 above)
Comment (cangtianhuang) [reply tester/accuracy.py]: done

**Review by Cutelemon6 — APPROVED** [review]
LGTM

**Review by wanghuancoder — APPROVED** [review]
LGTM
